### ISSUE TITLE
Nextcloud uploading files >64MB fails

### ISSUE THREAD

**Issue by kevinkk525** [storj/gateway#4]
This is basically the same issue as https://github.com/storj/storj/issues/3765 but that one was created with the old gateway so I thought I'll open a new ticket in this repository with the information from the current release rc6:

If i try to upload a file with a size bigger than the segment size of 64MB then the upload fails with this error message shown in Nextcloud GUI and the GUI just remains in the state of an unfinished upload:
```
Error when assembling chunks, status code 504
```
(This also happens with files smaller than 64MB but those still do get uploaded correctly and are there after a refresh of the gui)

According to netdata statistics the file still gets uploaded when the message is shown. However, after the upload is done, it seems like nextcloud tries to delete the file again and the gateway shows this warning:
```
2020-03-11T16:18:03.751+0100    WARN    Failed deleting object  {"path": "test-bucket/yEd-3.19.1.1_with-JRE13_64-bit_setup.exe", "error": "object not found: metainfo error: object not found: key not found: \"50e7dea3-3a19-417d-b7a5-57eff1d62a79/l/test-bucket/\\x02پk-S\\xafV\\xf5\\xa1\\xbe\\xab\\x92\\xd8\\x03\\xbbc!\\xab\\xaa\\xfd\\xc1\\x95 j\\xf5x\\xc8>C\\r\\x99\\b\\xaf\\xacD\\xcc\\xd3\\xc7Z\\xc8\\xf8LP\\xfe\\x02\\x9aެ\\v\\xb7\\xa87\\xaff_ݵ\\x89\\xcc\\x01\\x02\\xfa\\xfb\\x01\\x02\\x958P\\xf06\\xd9\"", "errorVerbose": "object not found: metainfo error: object not found: key not found: \"50e7dea3-3a19-417d-b7a5-57eff1d62a79/l/test-bucket/\\x02پk-S\\xafV\\xf5\\xa1\\xbe\\xab\\x92\\xd8\\x03\\xbbc!\\xab\\xaa\\xfd\\xc1\\x95 j\\xf5x\\xc8>C\\r\\x99\\b\\xaf\\xacD\\xcc\\xd3\\xc7Z\\xc8\\xf8LP\\xfe\\x02\\x9aެ\\v\\xb7\\xa87\\xaff_ݵ\\x89\\xcc\\x01\\x02\\xfa\\xfb\\x01\\x02\\x958P\\xf06\\xd9\"\n\tstorj.io/uplink/private/metainfo.(*Client).BeginDeleteObject:662\n\tstorj.io/uplink/private/storage/streams.(*streamStore).Delete:427\n\tstorj.io/uplink/private/storage/streams.(*streamStore).cancelHandler:531\n\tstorj.io/uplink/private/storage/streams.(*streamStore).Put.func1:106\n\tstorj.io/uplink/private/storage/streams.(*streamStore).Put:223\n\tstorj.io/uplink/private/storage/streams.(*shimStore).Put:54\n\tstorj.io/uplink/private/stream.NewUpload.func1:43\n\tgolang.org/x/sync/errgroup.(*Group).Go.func1:57"}
```

And nextcloud shows this error log:
```
[files_external] Error: Aws\S3\Exception\S3Exception: Error executing "CopyObject" on "http://localhost:7777/test-bucket/yEd-3.19.1.1_with-JRE13_64-bit_setup.exe"; AWS HTTP error: Client error: `PUT http://localhost:7777/test-bucket/yEd-3.19.1.1_with-JRE13_64-bit_setup.exe` resulted in a `404 Not Found` response:
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message> (truncated...)
 NoSuchKey (client): The specified key does not exist. - <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Key>yEd-3.19.1.1_with-JRE13_64-bit_setup.exe</Key><BucketName>test-bucket</BucketName><Resource>/test-bucket/yEd-3.19.1.1_with-JRE13_64-bit_setup.exe</Resource><RequestId>15FB48EEA0E4BA69</RequestId><HostId>fd7b4035-a365-4eea-9163-c9863d7967c8</HostId></Error> at <<closure>>

 0. /var/www/nextcloud/3rdparty/aws/aws-sdk-php/src/WrappedHttpHandler.php line 97
    Aws\WrappedHttpHandler->parseError("*** sensitive parameter replaced ***", GuzzleHttp\Psr7\Request {}, Aws\Command {}, [])
 1. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 203
    Aws\WrappedHttpHandler->Aws\{closure}("*** sensitive parameters replaced ***")
 2. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 174
    GuzzleHttp\Promise\Promise::callHandler(2, "*** sensitive parameter replaced ***", [GuzzleHttp\Prom ... }])
 3. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/RejectedPromise.php line 40
    GuzzleHttp\Promise\Promise::GuzzleHttp\Promise\{closure}("*** sensitive parameters replaced ***")
 4. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/TaskQueue.php line 47
    GuzzleHttp\Promise\RejectedPromise::GuzzleHttp\Promise\{closure}("*** sensitive parameters replaced ***")
 5. /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php line 98
    GuzzleHttp\Promise\TaskQueue->run()
 6. /var/www/nextcloud/3rdparty/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php line 125
    GuzzleHttp\Handler\CurlMultiHandler->tick()
 7. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 246
    GuzzleHttp\Handler\CurlMultiHandler->execute(true)
 8. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 223
    GuzzleHttp\Promise\Promise->invokeWaitFn()
 9. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 267
    GuzzleHttp\Promise\Promise->waitIfPending()
10. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 225
    GuzzleHttp\Promise\Promise->invokeWaitList()
11. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 267
    GuzzleHttp\Promise\Promise->waitIfPending()
12. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 225
    GuzzleHttp\Promise\Promise->invokeWaitList()
13. /var/www/nextcloud/3rdparty/guzzlehttp/promises/src/Promise.php line 62
    GuzzleHttp\Promise\Promise->waitIfPending()
14. /var/www/nextcloud/3rdparty/aws/aws-sdk-php/src/AwsClientTrait.php line 58
    GuzzleHttp\Promise\Promise->wait()
15. /var/www/nextcloud/3rdparty/aws/aws-sdk-php/src/AwsClientTrait.php line 77
    Aws\AwsClient->execute(Aws\Command {})
16. /var/www/nextcloud/apps/files_external/lib/Lib/Storage/AmazonS3.php line 572
    Aws\AwsClient->__call("copyObject", [{Bucket: "test- ... }])
17. /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php line 354
    OCA\Files_External\Lib\Storage\AmazonS3->touch("yEd-3.19.1.1_wi ... e", 1572713067)
18. /var/www/nextcloud/lib/private/Files/Storage/Wrapper/PermissionsMask.php line 100
    OC\Files\Storage\Wrapper\Wrapper->touch("yEd-3.19.1.1_wi ... e", 1572713067)
19. /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php line 354
    OC\Files\Storage\Wrapper\PermissionsMask->touch("yEd-3.19.1.1_wi ... e", 1572713067)
20. /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Availability.php line 364
    OC\Files\Storage\Wrapper\Wrapper->touch("yEd-3.19.1.1_wi ... e", 1572713067)
21. /var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php line 354
    OC\Files\Storage\Wrapper\Availability->touch("yEd-3.19.1.1_wi ... e", 1572713067)
22. /var/www/nextcloud/apps/files_accesscontrol/lib/StorageWrapper.php line 381
    OC\Files\Storage\Wrapper\Wrapper->touch("yEd-3.19.1.1_wi ... e", 1572713067)
23. /var/www/nextcloud/lib/private/Files/View.php line 1160
    OCA\FilesAccessControl\StorageWrapper->touch("yEd-3.19.1.1_wi ... e", 1572713067)
24. /var/www/nextcloud/lib/private/Files/View.php line 569
    OC\Files\View->basicOperation("touch", "/STORJS3/yEd-3. ... e", ["touch"], 1572713067)
25. /var/www/nextcloud/apps/dav/lib/Connector/Sabre/File.php line 302
    OC\Files\View->touch("/STORJS3/yEd-3. ... e", 1572713067)
26. /var/www/nextcloud/apps/dav/lib/Connector/Sabre/Directory.php line 156
    OCA\DAV\Connector\Sabre\File->put(null)
27. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Tree.php line 316
    OCA\DAV\Connector\Sabre\Directory->createFile("yEd-3.19.1.1_wi ... e", null)
28. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Tree.php line 130
    Sabre\DAV\Tree->copyNode(OCA\DAV\Upload\FutureFile {}, OCA\DAV\Connector\Sabre\Directory {}, "yEd-3.19.1.1_wi ... e")
29. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Tree.php line 161
    Sabre\DAV\Tree->copy("uploads/kevin_k ... e", "files/kevin_koe ... e")
30. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php line 642
    Sabre\DAV\Tree->move("uploads/kevin_k ... e", "files/kevin_koe ... e")
31. <<closure>>
    Sabre\DAV\CorePlugin->httpMove(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
32. /var/www/nextcloud/3rdparty/sabre/event/lib/EventEmitterTrait.php line 105
    call_user_func_array([Sabre\DAV\CorePlugin {},"httpMove"], [Sabre\HTTP\Requ ... }])
33. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 479
    Sabre\Event\EventEmitter->emit("method:MOVE", [Sabre\HTTP\Requ ... }])
34. /var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php line 254
    Sabre\DAV\Server->invokeMethod(Sabre\HTTP\Reque ... "}, Sabre\HTTP\Response {})
35. /var/www/nextcloud/apps/dav/lib/Server.php line 319
    Sabre\DAV\Server->exec()
36. /var/www/nextcloud/apps/dav/appinfo/v2/remote.php line 35
    OCA\DAV\Server->exec()
37. /var/www/nextcloud/remote.php line 165
    require_once("/var/www/nextcl ... p")

MOVE /remote.php/dav/uploads/kevin_koeck/web-file-upload-59495c93aab61c15d58f986efdb146fe-1583939769325/.file
from <ip> by kevin_koeck at 2020-03-11T15:18:03+00:00
```

This seems to be slightly different than how it was with the old gateway. 
Now there is no file in the bucket, while with the old gateway there was a file of 0B instead of the original size.

But since the original issue in storj/storj is tracked in your internal jira, you might already be aware of this current state.

**Comment by kaloyan-raev** [comment]
@kevinkk525 I spend some time today reproducing and investigating the issue. It seems there are two issues:
1. For some reason, after uploading the object, NextCloud requests `CopyObject` with the same source and destination, i.e. it tries to overwrite the object over itself. I think it tries to encrypt the uploaded object with S3 server-side encryption. The S3 Gateway didn't handle this situation well, which resulted in destroying the object, especially if it was a big one.
2. Larger uploads seem to take longer time than the PHP max execution time, which causes the `Error when assembling chunks` error.

We released gateway v1.0.4 to resolve the first issue: https://github.com/storj/gateway/releases/tag/v1.0.4

For the second issue, you should take a look at how to increase the max execution time in your PHP configuration. Here is an article that talks about this: https://help.nextcloud.com/t/error-when-assembling-chunks-status-code-504/56751

**Comment by kevinkk525** [comment]
Thank you! This version works great! No problems uploading multiple big files.
The gateway is really useful now :)

Thanks for the link about the issue 2, that is indeed on the nextcloud side only.
